File: CS/CodeUtils/CodeExampleModel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using DevExpress.XtraRichEdit;
using DevExpress.XtraTreeList;

namespace RichEditDocumentServerAPIExample.CodeUtils
{
    public class RichEditExample : RichEditNode
    {
        public RichEditExample(string name, string codeCS, string codeVB, Action<RichEditDocumentServer> action, bool saveResult) : base(name)
        {
            Action = action;
            CodeCS = codeCS;
            CodeVB = codeVB;
            SaveResult = saveResult;
        }
        public Action<RichEditDocumentServer> Action { get; private set; }
        public string CodeCS { get; set; }
        public string CodeVB { get; set; }
        public bool SaveResult { get; set; }

    }
    public class RichEditNode
    {
        GroupsOfRichEditExamples groups = new GroupsOfRichEditExamples();

        public RichEditNode(string name)
        {
            Name = name;
        }
        [Browsable(false)]
        public GroupsOfRichEditExamples Groups { get { return groups; } }
        public string Name { get; set; }
        [Browsable(false)]
        public GroupsOfRichEditExamples Owner { get; set; }
    }

    public class GroupsOfRichEditExamples : BindingList<RichEditNode>, TreeList.IVirtualTreeListData
    {
        void TreeList.IVirtualTreeListData.VirtualTreeGetChildNodes(VirtualTreeGetChildNodesInfo info)
        {
            RichEditNode obj = info.Node as RichEditNode;
            info.Children = obj.Groups;
        }
        protected override void InsertItem(int index, RichEditNode item)
        {
            item.Owner = this;
            base.InsertItem(index, item);
        }
        void TreeList.IVirtualTreeListData.VirtualTreeGetCellValue(VirtualTreeGetCellValueInfo info)
        {
            RichEditNode obj = info.Node as RichEditNode;
            switch (info.Column.Caption)
            {
                case "Name":
                    info.CellData = obj.Name;
                    break;
            }
        }
        void TreeList.IVirtualTreeListData.VirtualTreeSetCellValue(VirtualTreeSetCellValueInfo info)
        { }
    }

    public class CodeExampleGroup
    {
        public string Name { get; set; }
        public List<CodeExample> Examples { get; set; }
        public int Id { get; set; }
    }

    public class CodeExample
    {
        public string CodeCS { get; set; }
        //public string CodeCsHelper { get; set; }
        public string CodeVB { get; set; }
        //public string CodeVbHelper { get; set; }
        public string RegionName { get; set; }
        public string HumanReadableGroupName { get; set; }
        public string ExampleGroup { get; set; }
        public int Id { get; set; }
    }

    public enum ExampleLanguage
    {
        Csharp = 0,
        VB = 1
    }
}

File: VB/CodeUtils/CodeExampleModel.vb
Imports System.Collections.Generic

Namespace RichEditDocumentServerAPIExample.CodeUtils

    Public Class CodeExampleGroup

        Public Sub New()
        End Sub

        Public Property Name As String

        Public Property Examples As List(Of CodeExample)

        Public Property Id As Integer
    End Class

    Public Class CodeExample

        Public Property CodeCS As String

        Public Property CodeCsHelper As String

        Public Property CodeVB As String

        Public Property CodeVbHelper As String

        Public Property RegionName As String

        Public Property HumanReadableGroupName As String

        Public Property ExampleGroup As String

        Public Property Id As Integer
    End Class

    Public Class CodeExampleCollection
        Inherits List(Of CodeExample)

        Public Sub Merge(ByVal example As CodeExample)
            Dim item As CodeExample = Find(Function(x) x.HumanReadableGroupName.Equals(example.HumanReadableGroupName) AndAlso x.RegionName.Equals(example.RegionName))
            If item Is Nothing Then
                item = New CodeExample()
                item.HumanReadableGroupName = example.HumanReadableGroupName
                item.RegionName = example.RegionName
                Add(item)
            End If

            item.CodeCS += example.CodeCS
            item.CodeCsHelper += example.CodeCsHelper
            item.CodeVB += example.CodeVB
            item.CodeVbHelper += example.CodeVbHelper
        End Sub

        Public Sub Merge(ByVal exampleList As List(Of CodeExample))
            For Each item As CodeExample In exampleList
                Merge(item)
            Next
        End Sub
    End Class

    Public Enum ExampleLanguage
        Csharp = 0
        VB = 1
    End Enum
End Namespace
